File: server/src/services/kalshi.js
/**
 * Kalshi API Service
 * Fetches live prediction market data from Kalshi
 */
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

class KalshiClient {
  constructor() {
    this.apiKey = process.env.KALSHI_API_KEY;
    this.privateKeyPath = process.env.KALSHI_PRIVATE_KEY_PATH || path.join(__dirname, '../../kalshi_private_key.pem');
    this.baseUrl = process.env.KALSHI_API_URL || 'https://api.elections.kalshi.com/trade-api/v2';
    this.privateKey = this.loadPrivateKey();
  }

  loadPrivateKey() {
    try {
      // Try multiple paths
      const paths = [
        this.privateKeyPath,
        path.join(__dirname, '../../kalshi_private_key.pem'),
        './kalshi_private_key.pem',
        '/app/kalshi_private_key.pem'
      ];
      
      for (const p of paths) {
        try {
          if (fs.existsSync(p)) {
            return fs.readFileSync(p, 'utf8');
          }
        } catch {}
      }
      
      // Try from environment variable (raw PEM or base64 encoded)
      if (process.env.KALSHI_API_SECRET) {
        const secret = process.env.KALSHI_API_SECRET.trim();
        if (secret.includes('BEGIN')) return secret;
        return Buffer.from(secret, 'base64').toString('utf8');
      }
      if (process.env.KALSHI_PRIVATE_KEY) {
        const secret = process.env.KALSHI_PRIVATE_KEY.trim();
        if (secret.includes('BEGIN')) return secret;
        return Buffer.from(secret, 'base64').toString('utf8');
      }
      
      console.warn('Kalshi private key not found');
      return null;
    } catch (error) {
      console.error('Error loading Kalshi private key:', error);
      return null;
    }
  }

  signRequest(timestamp, method, path) {
    if (!this.privateKey) return '';
    
    try {
      const message = `${timestamp}${method}${path}`;
      const sign = crypto.createSign('RSA-SHA256');
      sign.update(message);
      sign.end();
      return sign.sign({
        key: this.privateKey,
        padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
        saltLength: crypto.constants.RSA_PSS_SALTLEN_DIGEST,
      }, 'base64');
    } catch (error) {
      console.error('Error signing request:', error);
      return '';
    }
  }

  getHeaders(method, path) {
    const timestamp = Date.now().toString();
    const signature = this.signRequest(timestamp, method, path);
    
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    };

    if (this.apiKey) {
      headers['KALSHI-ACCESS-KEY'] = this.apiKey;
    }
    if (signature) {
      headers['KALSHI-ACCESS-SIGNATURE'] = signature;
      headers['KALSHI-ACCESS-TIMESTAMP'] = timestamp;
    }

    return headers;
  }

  async getMarkets(limit = 100, status = 'open') {
    const path = `/markets?limit=${limit}&status=${status}`;
    const url = `${this.baseUrl}${path}`;
    
    try {
      const signPath = new URL(url).pathname;
      const headers = this.privateKey ? this.getHeaders('GET', signPath) : { 'Accept': 'application/json' };
      const response = await fetch(url, { headers, timeout: 10000 });
      
      if (response.ok) {
        const data = await response.json();
        return data.markets || [];
      } else {
        console.error(`Kalshi API error: ${response.status}`);
        return [];
      }
    } catch (error) {
      console.error('Error fetching Kalshi markets:', error);
      return [];
    }
  }

  async getEvents(limit = 50, status = 'open') {
    const path = `/events?limit=${limit}&status=${status}`;
    const url = `${this.baseUrl}${path}`;
    
    try {
      const signPath = new URL(url).pathname;
      const headers = this.privateKey ? this.getHeaders('GET', signPath) : { 'Accept': 'application/json' };
      const response = await fetch(url, { headers, timeout: 10000 });
      
      if (response.ok) {
        const data = await response.json();
        return data.events || [];
      } else {
        console.error(`Kalshi API error: ${response.status}`);
        return [];
      }
    } catch (error) {
      console.error('Error fetching Kalshi events:', error);
      return [];
    }
  }

  formatMarketForArb(market) {
    const yesPrice = (market.yes_ask || market.last_price || 50) / 100;
    const noPrice = (market.no_ask || (100 - (market.yes_ask || market.last_price || 50))) / 100;
    
    return {
      id: market.ticker,
      event: market.title || market.ticker,
      category: market.category || 'Other',
      kalshi: {
        yes: yesPrice,
        no: noPrice,
        volume: market.volume || 0,
        openInterest: market.open_interest || 0
      },
      expiry: market.close_time || market.expiration_time,
      status: market.status || 'open'
    };
  }
}

export const kalshiClient = new KalshiClient();

export async function getKalshiMarkets(limit = 50) {
  const markets = await kalshiClient.getMarkets(limit);
  return markets.map(m => kalshiClient.formatMarketForArb(m));
}

export async function getKalshiEvents(limit = 30) {
  return kalshiClient.getEvents(limit);
}

export async function getArbitrageOpportunities(limit = 20) {
  const kalshiMarkets = await getKalshiMarkets(100);
  const opportunities = [];
  
  for (const market of kalshiMarkets.slice(0, limit)) {
    const kalshiYes = market.kalshi?.yes || 0.5;
    
    // Simulate Polymarket spread (in production, fetch from Polymarket API)
    const polySpread = (Math.random() - 0.5) * 0.1;
    const polyYes = Math.max(0.01, Math.min(0.99, kalshiYes + polySpread));
    
    const spread = Math.abs(kalshiYes - polyYes) * 100;
    
    if (spread >= 2.0) {
      opportunities.push({
        id: market.id,
        event: market.event,
        category: market.category,
        kalshi: {
          yes: Math.round(kalshiYes * 100) / 100,
          volume: market.kalshi?.volume || 0
        },
        polymarket: {
          yes: Math.round(polyYes * 100) / 100,
          volume: `${Math.floor(Math.random() * 5000 + 100)}K`
        },
        spread: Math.round(spread * 10) / 10,
        profit: `$${Math.round(spread * 10)} per $1000`,
        confidence: spread >= 4 ? 'High' : spread >= 3 ? 'Medium' : 'Low',
        expiry: market.expiry
      });
    }
  }
  
  // Sort by spread (highest first)
  opportunities.sort((a, b) => b.spread - a.spread);
  
  return opportunities.slice(0, limit);
}

File: backend/api/kalshi.py
"""
Kalshi API Routes
Endpoints for prediction market data
"""
from fastapi import APIRouter, HTTPException, Query
from typing import List, Dict, Any, Optional
from services.kalshi_service import get_kalshi_markets, get_kalshi_events, kalshi_client

router = APIRouter(prefix="/api/v1/kalshi", tags=["kalshi"])


@router.get("/markets")
async def get_markets(
    limit: int = Query(50, ge=1, le=200),
    category: Optional[str] = None
) -> Dict[str, Any]:
    """
    Get live Kalshi prediction markets
    
    Returns markets with current prices, volume, and metadata
    """
    try:
        markets = get_kalshi_markets(limit=limit)
        
        # Filter by category if specified
        if category and category != "All":
            markets = [m for m in markets if m.get("category", "").lower() == category.lower()]
        
        return {
            "success": True,
            "count": len(markets),
            "markets": markets
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching Kalshi markets: {str(e)}")


@router.get("/markets/{ticker}")
async def get_market(ticker: str) -> Dict[str, Any]:
    """
    Get a specific Kalshi market by ticker
    """
    try:
        market = kalshi_client.get_market(ticker)
        if not market:
            raise HTTPException(status_code=404, detail=f"Market {ticker} not found")
        
        return {
            "success": True,
            "market": kalshi_client.format_market_for_arb(market)
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching market: {str(e)}")


@router.get("/events")
async def get_events(
    limit: int = Query(30, ge=1, le=100)
) -> Dict[str, Any]:
    """
    Get Kalshi events (market categories/groupings)
    """
    try:
        events = get_kalshi_events(limit=limit)
        return {
            "success": True,
            "count": len(events),
            "events": events
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching events: {str(e)}")


@router.get("/arbitrage")
async def get_arbitrage_opportunities(
    limit: int = Query(20, ge=1, le=50)
) -> Dict[str, Any]:
    """
    Get potential arbitrage opportunities
    
    Compares Kalshi prices with known Polymarket prices
    to find spreads that could be profitable
    """
    try:
        kalshi_markets = get_kalshi_markets(limit=100)
        
        # Mock Polymarket data for comparison (in production, fetch from Polymarket API)
        # Polymarket uses CLOB system - we'd need to query their GraphQL API
        opportunities = []
        
        for market in kalshi_markets[:limit]:
            kalshi_yes = market.get("kalshi", {}).get("yes", 0.5)
            
            # Simulate Polymarket price (in production, fetch real data)
            # Adding small random spread for demo
            import random
            poly_spread = random.uniform(-0.05, 0.05)
            poly_yes = max(0.01, min(0.99, kalshi_yes + poly_spread))
            
            spread = abs(kalshi_yes - poly_yes) * 100
            
            if spread >= 2.0:  # Only show opportunities with 2%+ spread
                opportunities.append({
                    "id": market.get("id"),
                    "event": market.get("event"),
                    "category": market.get("category"),
                    "kalshi": {
                        "yes": round(kalshi_yes, 2),
                        "volume": market.get("kalshi", {}).get("volume", 0)
                    },
                    "polymarket": {
                        "yes": round(poly_yes, 2),
                        "volume": f"{random.randint(100, 5000)}K"  # Mock for now
                    },
                    "spread": round(spread, 1),
                    "profit": f"${int(spread * 10)} per $1000",
                    "confidence": "High" if spread >= 4 else "Medium" if spread >= 3 else "Low",
                    "expiry": market.get("expiry")
                })
        
        # Sort by spread (highest first)
        opportunities.sort(key=lambda x: x["spread"], reverse=True)
        
        return {
            "success": True,
            "count": len(opportunities),
            "opportunities": opportunities[:limit]
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error calculating arbitrage: {str(e)}")


@router.get("/health")
async def health_check() -> Dict[str, Any]:
    """Check if Kalshi API connection is working"""
    try:
        markets = kalshi_client.get_markets(limit=1)
        connected = len(markets) > 0
        
        return {
            "success": True,
            "kalshi_connected": connected,
            "api_key_set": bool(kalshi_client.api_key),
            "private_key_loaded": bool(kalshi_client.private_key)
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "kalshi_connected": False
        }
